text/0626-invoke-helper.md

```diff
@@ -142,124 +118,252 @@ function myProgram(data = []) {
+   templating, they're just being used for logic effectively. So, there's no
+   reason to add the overhead of a component here.
+
+3. We could use other template constructs, such as helpers or modifiers. Both
+   helpers and modifiers have lifecycles, like components, and can be used to
+   contain functionality. Modifiers aren't really a good choice here though,
+   because it would require us to add an element that we don't need. So, helpers
+   are the better option.
+
+   Helpers work, but like components they require us to move some of our logic
+   into the template, even if that isn't really necessary:
+
+    ```js
+    // app/helpers/doc-title.js
+    import { helper } from '@ember/component/helper';
+
+    export default helper(([title]) => {
+      document.title = title;
+    });
+    ```
+    ```js
+    // app/helpers/fetch-data.js
+    import Helper from '@ember/component/helper';
+    import { fetch, cancelFetch } from 'fetch-with-cancel';
+
+    export default class FetchData extends Helper {
+      @tracked data;
+
+      compute([url]) {
+        if (this._url !== url) {
+          this.url = url;
+
+          // Cancel the previous fetch if it's still running
+          cancelFetch(this.promise);
+
+          // create a new fetch request, and set the data property to the response
+          this.promise = fetch(url)
+            .then((response) => response.json());
+            .then((data) => this.data = data);
+        }
+
+        return this.data;
+      }
+
+      willDestroy() {
+        cancelFetch(this.promise);
+      }
+    }
+    ```
+    ```hbs
+    <!-- app/components/search.hbs -->
+    {{doc-title 'Search Result for "{{@query}}"'}}
+
+    {{#let (fetch-data "www.example.com/search?query={{@query}}") as |data|}}
+      ...
+    {{/let}}
+    ```
+
+Out of these options, helpers are the closest to what we want - they are low
+overhead, they produce computed values directly without a template, and
+with the recent addition of effect helpers they can be used side-effect to
+accomplish tasks like setting the document title. The only downside is that they
+can only be invoked in templates, so they require you to design your components
+around using them in templates only. This can be difficult to do in many cases,
+where the data wants to be accessed to create derived state for instance.
+
+This RFC proposes adding a way to create helpers within JavaScript directly,
+extending the reactive model in a way that allows users to extract common
+reactive code and patterns, and reuse them transparently. This will make helpers
+the new reactive atom of the system - the reactive equivalent of a "function" in
+our incremental model. Like components, they have a lifecycle, and can update
+over time. Unlike components, they can exist nested in JavaScript classes _and_
+in templates, and they can produce any type of value, making them much more
+flexible.
```

Comment:
I'm not sure I'm following this part, is the point you are making that components can't be nested in JavaScript?

Reply:
Yes, and that makes them less flexible.

Reply:
Yes, and that makes them less flexible.

Reply:
Thank you for clarifying! From my perspective _some_ of this concern is mitigated with upcoming changes to allow components to be "just values", but still seems good to state this here (I was just confused).